pkg/controller/direct/iam/partialpolicy_controller.go

```diff
@@ -44,7 +45,42 @@ const (
 )
 
 func init() {
-	registry.RegisterModel(krm.IAMPartialPolicyGVK, NewIAMPartialPolicyModel)
+	rg := &IAMPartialPolicyReconcileGate{}
+	registry.RegisterModelWithReconcileGate(krm.IAMPartialPolicyGVK, NewIAMPartialPolicyModel, rg)
+}
+
+// IAMPartialPolicyReconcileGate implements a reconcile gate for IAMPartialPolicy
+// that defaults to the direct controller for better performance, but allows
+// fallback to the handwritten controller when needed.
+//
+// By default, IAMPartialPolicy resources use the direct controller.
+// To opt out and use the handwritten controller:
+//   - alpha.cnrm.cloud.google.com/reconciler: "legacy"
+//
+// To explicitly opt in to the direct controller:
+//   - alpha.cnrm.cloud.google.com/reconciler: "direct"
+type IAMPartialPolicyReconcileGate struct {
+	optIn predicate.OptInToDirectReconciliation
+}
+
+var _ predicate.ReconcileGate = &IAMPartialPolicyReconcileGate{}
+
+func (r *IAMPartialPolicyReconcileGate) ShouldReconcile(o *unstructured.Unstructured) bool {
+	// Check for alpha reconciler annotation to opt out
+	if v := o.GetAnnotations()["alpha.cnrm.cloud.google.com/reconciler"]; v == "legacy" {
+		return false
+	}
+
+	// todo acpana integrate with experiments controls for CCC/ CC
```

Comment:
Thank you for clarifying this. If this integration hasn't be supported, then I suggest we don't merge #4930, as the feature won't work.

Reply:
> Just curious, can users opt-in using direct controller for IAMPartialPolicy right now?

Yes it's on by default with this change